### PR TITLE
Update Samplers to conform to Spec

### LINKED ISSUE #519: Sampler API must accept Attributes and Links and return Attributes
While looking at options for addressing #496 I found that the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-tracing.md#sampling) already requires that samplers receive the initial set of Attributes and Links attached to a span.  It also requires that the `SamplingResult` include a property for Attributes related to the sampling decision and that the `Decision` be one of three options in an Enum and not a bool.  The current Go API and SDK trace packages are non-conformant to varying degrees.

### PR DESCRIPTION
**Pull request by Aneurysm9**
Fixes #519.

### CODE REVIEW

File: api/trace/sampler.go [hunk 1]
```diff
@@ -26,6 +26,9 @@ type Sampler interface {
 		traceID core.TraceID,
 		spanID uint64,
```
Comment (jmacd) [inline api/trace/sampler.go]: This uses `core.SpanID` elsewhere. Would you mind fixing this up?

File: sdk/trace/sampling.go [hunk 2]
```diff
@@ -30,11 +35,46 @@ type SamplingParameters struct {
...
+}
+
+// SamplingDecision indicates whether a span is recorded and sampled.
+type SamplingDecision uint8
+
+// Valid sampling decisions
+const (
+	NotRecord SamplingDecision = iota
+	Record
+	RecordAndSampled
```
Comment (krnowak) [inline sdk/trace/sampling.go]: I'd say it should be "RecordedAndSampled" or "RecordAndSample", but I see that the spec specified the decision name like that…

File: sdk/trace/sampling.go [hunk 3]
```diff
@@ -49,37 +89,53 @@ func ProbabilitySampler(fraction float64) Sampler {
...
-	return func(p SamplingParameters) SamplingDecision {
-		if p.ParentContext.IsSampled() {
-			return SamplingDecision{Sample: true}
-		}
-		x := binary.BigEndian.Uint64(p.TraceID[0:8]) >> 1
-		return SamplingDecision{Sample: x < traceIDUpperBound}
+
+	return &probabilitySampler{
+		traceIDUpperBound: uint64(fraction * (1 << 63)),
+		description:       fmt.Sprintf("ProbabilitySampler{%v}", fraction),
```
Comment (krnowak) [inline sdk/trace/sampling.go]:
```suggestion
		description:       fmt.Sprintf("ProbabilitySampler{%g}", fraction),
```

File: sdk/trace/sampling.go [hunk 4]
```diff
@@ -49,37 +89,53 @@ func ProbabilitySampler(fraction float64) Sampler {
...
 	}
 }
 
+type alwaysOnSampler bool
```
Comment (krnowak) [inline sdk/trace/sampling.go]: For empty types, you usually use `struct{}` in Go.

```suggestion
type alwaysOnSampler struct{}
```

File: sdk/trace/sampling.go [hunk 5]
```diff
@@ -49,37 +89,53 @@ func ProbabilitySampler(fraction float64) Sampler {
...
+
 // AlwaysSample returns a Sampler that samples every trace.
 // Be careful about using this sampler in a production application with
 // significant traffic: a new trace will be started and exported for every
 // request.
 func AlwaysSample() Sampler {
-	return func(p SamplingParameters) SamplingDecision {
-		return SamplingDecision{Sample: true}
-	}
+	return alwaysOnSampler(true)
```
Comment (krnowak) [inline sdk/trace/sampling.go]: With the above change, this would become:
```suggestion
	return alwaysOnSampler{}
```

File: sdk/trace/sampling.go [hunk 6]
```diff
@@ -49,37 +89,53 @@ func ProbabilitySampler(fraction float64) Sampler {
...
+}
+
+type alwaysOffSampler bool
```
Comment (krnowak) [inline sdk/trace/sampling.go]: Same here.
```suggestion
type alwaysOffSampler struct{}
```

File: sdk/trace/sampling.go [hunk 7]
```diff
@@ -49,37 +89,53 @@ func ProbabilitySampler(fraction float64) Sampler {
...
+func (as alwaysOffSampler) Description() string {
+	return "AlwaysOffSampler"
 }
 
 // NeverSample returns a Sampler that samples no traces.
 func NeverSample() Sampler {
-	return func(p SamplingParameters) SamplingDecision {
-		return SamplingDecision{Sample: false}
-	}
+	return alwaysOffSampler(false)
```
Comment (krnowak) [inline sdk/trace/sampling.go]:
```suggestion
	return alwaysOffSampler{}
```

File: sdk/trace/sampling_test.go [hunk 8]
```diff
@@ -30,7 +30,7 @@ func TestAlwaysParentSampleWithParentSampled(t *testing.T) {
 		SpanID:     spanID,
 		TraceFlags: core.TraceFlagsSampled,
 	}
-	if !sampler(sdktrace.SamplingParameters{ParentContext: parentCtx}).Sample {
+	if sampler.ShouldSample(sdktrace.SamplingParameters{ParentContext: parentCtx}).Decision != sdktrace.RecordAndSampled {
 		t.Error("Sampling decision should be true")
```
Comment (krnowak) [inline sdk/trace/sampling_test.go]: Could you fix this message too? Thanks.

File: sdk/trace/sampling_test.go [hunk 9]
```diff
@@ -43,7 +43,7 @@ func TestAlwaysParentSampleWithParentNotSampled(t *testing.T) {
 		TraceID: traceID,
 		SpanID:  spanID,
 	}
-	if sampler(sdktrace.SamplingParameters{ParentContext: parentCtx}).Sample {
+	if sampler.ShouldSample(sdktrace.SamplingParameters{ParentContext: parentCtx}).Decision != sdktrace.NotRecord {
 		t.Error("Sampling decision should be false")
```
Comment (krnowak) [inline sdk/trace/sampling_test.go]: Could you fix this message? Thanks.

File: sdk/trace/sampling.go (hunk 2 above)
Comment (Aneurysm9) [reply sdk/trace/sampling.go]: Yeah, I'm not a fan either, but I think sticking to the spec is the thing to do here.  At least until someone changes the spec :)

File: api/trace/sampler.go (hunk 1 above)
Comment (Aneurysm9) [reply api/trace/sampler.go]: Done.

**Comment by Aneurysm9** [comment]
@jmacd and @krnowak thanks for the feedback.  I've addressed the requested changes.

**Pull request merged**